library/std/src/sys/unix/os.rs

```diff
@@ -368,21 +368,12 @@ pub fn current_exe() -> io::Result<PathBuf> {
 
 #[cfg(any(target_os = "solaris", target_os = "illumos"))]
 pub fn current_exe() -> io::Result<PathBuf> {
-    extern "C" {
-        fn getexecname() -> *const c_char;
-    }
-    unsafe {
-        let path = getexecname();
-        if path.is_null() {
-            Err(io::Error::last_os_error())
-        } else {
-            let filename = CStr::from_ptr(path).to_bytes();
-            let path = PathBuf::from(<OsStr as OsStrExt>::from_bytes(filename));
-
-            // Prepend a current working directory to the path if
-            // it doesn't contain an absolute pathname.
-            if filename[0] == b'/' { Ok(path) } else { getcwd().map(|cwd| cwd.join(path)) }
-        }
+    match crate::fs::read_link("/proc/self/path/a.out") {
+        Err(ref e) if e.kind() == io::ErrorKind::NotFound => Err(io::Error::new_const(
+            io::ErrorKind::Uncategorized,
+            &"no /proc/self/a.out available. Is /proc mounted?",
```

Comment:
Maybe still use `getexecname` if `/proc` is not mounted? I am guessing this could happen inside containers.

Reply:
fair point